mlflow/pyfunc/__init__.py

```diff
@@ -656,15 +674,285 @@ def _warn_dependency_requirement_mismatches(model_path):
+    """
+    We need create _RestoredPythonEnvSubprocessMgr instance
+    from spark application driver side.
+    """
+    def __init__(self, py_version, requirements_file_path, loader_module, model_path):
+        self._py_version = py_version
+        self._requirements_file_path = requirements_file_path
+        self._loader_module = loader_module
+        self._model_path = model_path
+
+        self._subproc = None
+        self._python_installation_dir = None
+        self._virtual_env_dir = None
+
+        self._virtual_env_python_bin = None
+        self._comm_sock = None
+        self._comm_stream = None
+
+    _PYTHON_ENV_CACHE_DIR = None
+
+    @staticmethod
+    def _get_or_create_python_env_cache_dir():
+        if _RestoredPythonEnvSubprocessMgr._PYTHON_ENV_CACHE_DIR is not None:
+            return _RestoredPythonEnvSubprocessMgr._PYTHON_ENV_CACHE_DIR
+
+        if databricks_utils.is_in_databricks_runtime():
+            # In databricks, the '/local_disk0/.ephemeral_nfs' is mounted as NFS disk
+            # the data stored in the disk is shared with all remote nodes.
+            root_dir = os.path.join(_DATABRICKS_NFS_CACHE_ROOT_DIR, "env")
+            os.makedirs(root_dir, exist_ok=True)
+            _RestoredPythonEnvSubprocessMgr._PYTHON_ENV_CACHE_DIR = tempfile.mkdtemp(dir=root_dir)
+            # TODO: register deleting cache dir handler when exit
+        else:
+            """
+            import atexit
+            import shutil
+            _RestoredPythonEnvSubprocessMgr._PYTHON_ENV_CACHE_DIR = tempfile.mkdtemp()
+            atexit.register(shutil.rmtree, _RestoredPythonEnvSubprocessMgr._PYTHON_ENV_CACHE_DIR, ignore_errors=True)
+            """
+            _RestoredPythonEnvSubprocessMgr._PYTHON_ENV_CACHE_DIR = '/tmp'  # For testing purpose
+
+        os.makedirs(os.path.join(_RestoredPythonEnvSubprocessMgr._PYTHON_ENV_CACHE_DIR, 'pythons'), exist_ok=True)
+        os.makedirs(os.path.join(_RestoredPythonEnvSubprocessMgr._PYTHON_ENV_CACHE_DIR, 'venvs'), exist_ok=True)
+
+        return _RestoredPythonEnvSubprocessMgr._PYTHON_ENV_CACHE_DIR
+
+    @property
+    def comm_stream(self):
+        return self._comm_stream
+
+    @property
+    def subproc(self):
+        return self._subproc
+
+    def setup_restored_python_env(self):
+        env_cache_dir = _RestoredPythonEnvSubprocessMgr._get_or_create_python_env_cache_dir()
+        pyenv_home_path = os.path.join(
+            env_cache_dir, 'pythons'
+        )
+        self._python_installation_dir = os.path.join(
+            pyenv_home_path, '.pyenv', 'versions', self._py_version
+        )
+        if not os.path.exists(self._python_installation_dir):
+            # install specified python version
+            sub_env = os.environ.copy()
+            sub_env['HOME'] = pyenv_home_path
+            subprocess.run(
+                [_PYENV_CMD, 'install', '-v', self._py_version],
+                check=True, stdout=sys.stdout, stderr=sys.stderr,
```

Comment:
will this subprocess sys.stdout and sys.stderr be visible on the driver for Spark? 
Should we have this install information piped to a log file so that users don't get confused if seeing conflicting versions of libraries that are not actually installed in their session repl?

Reply:
Yes I will improve how to handle the subprocess output. 
redirect them to parent output is easier for user debugging, we can prepend string like "subprocess output: " to the output logs.